src/Servers/Kestrel/Core/src/Internal/Http/Http1ContentLengthMessageBody.cs

```diff
@@ -70,44 +70,46 @@ public override async ValueTask<ReadResult> ReadAsyncInternal(CancellationToken
                 }
                 catch (ConnectionAbortedException ex)
                 {
+                    _isReading = false;
                     throw new TaskCanceledException("The request was aborted", ex);
                 }
 
+                void ResetReadingState()
+                {
+                    _isReading = false;
+                    // Reset the timing read here for the next call to read.
+                    StopTimingRead(0);
+                    _context.Input.AdvanceTo(_readResult.Buffer.Start);
+                }
+
                 if (_context.RequestTimedOut)
                 {
+                    ResetReadingState();
                     BadHttpRequestException.Throw(RequestRejectionReason.RequestBodyTimeout);
                 }
 
-                // Make sure to handle when this is canceled here.
-                if (_readResult.IsCanceled)
+                if (_readResult.IsCompleted)
                 {
-                    if (Interlocked.Exchange(ref _userCanceled, 0) == 1)
-                    {
-                        // Ignore the readResult if it wasn't by the user.
-                        CreateReadResultFromConnectionReadResult();
-
-                        break;
-                    }
-                    else
-                    {
-                        // Reset the timing read here for the next call to read.
-                        StopTimingRead(0);
-                        continue;
-                    }
+                    ResetReadingState();
+                    ThrowUnexpectedEndOfRequestContent();
                 }
 
-                var readableBuffer = _readResult.Buffer;
-                var readableBufferLength = readableBuffer.Length;
-                StopTimingRead(readableBufferLength);
-
-                CheckCompletedReadResult(_readResult);
-
-                if (readableBufferLength > 0)
+                // Ignore the canceled readResult if it wasn't canceled by the user.
+                if ((!_readResult.IsCanceled && _readResult.Buffer.Length > 0) ||
+                    (_readResult.IsCanceled && Interlocked.Exchange(ref _userCanceled, 0) == 1))
                 {
                     CreateReadResultFromConnectionReadResult();
+                    StopTimingRead(_readResult.Buffer.Length);
```

Comment:
Note to self: don't recompute the buffer length.

Reply:
Where are you computing the buffer length here?

Reply:
We were before too. It was just stored into a local before so it wasn't recomputed.